presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,373 @@
+            Set<VariableReferenceExpression> joinKeys = context.get().getJoinKeys();
+
+            if (joinKeys.size() == 0 || context.get().getNumJoins() < 2) {
+                return scanNode;
+            }
+
+            System.out.println("Join keys: " + joinKeys);
+
+            if (!ImmutableSet.copyOf(scanNode.getOutputVariables()).containsAll(joinKeys)) {
+                // not all join keys are in the table scan: abort rewrite
+                return scanNode;
+            }
+
+            List<VariableReferenceExpression> groupingKeys = joinKeys.stream().collect(Collectors.toList());
+
+            TableLayoutResult layout = metadata.getLayout(
+                    session,
+                    scanNode.getTable(),
+                    Constraint.alwaysTrue(),
+                    Optional.of(scanNode.getOutputVariables().stream()
+                            .map(variable -> scanNode.getAssignments().get(variable))
+                            .collect(toImmutableSet())));
+
+            Map<VariableReferenceExpression, ColumnHandle> assignments = scanNode.getAssignments();
+
+            AggregationNode agg = new AggregationNode(
+                    scanNode.getSourceLocation(),
+                    planNodeIdAllocator.getNextId(),
+                    scanNode,
+                    ImmutableMap.of(),
+                    singleGroupingSet(groupingKeys),
+                    ImmutableList.of(),
+                    SINGLE,
+                    Optional.empty(),
+                    Optional.empty());
+
+            ImmutableList.Builder<VariableReferenceExpression> outputVariablesBuilder = ImmutableList.builder();
+            ImmutableList.Builder<JoinNode.EquiJoinClause> equiJoinClausesBuilder = ImmutableList.builder();
+
+            ImmutableMap.Builder<VariableReferenceExpression, ColumnHandle> assignmentsBuilder = ImmutableMap.builder();
+
+            for (VariableReferenceExpression var : scanNode.getOutputVariables()) {
+                if (joinKeys.contains(var)) {
+                    VariableReferenceExpression newVar = planVariableAllocator.newVariable(var.getName(), var.getType());
+                    outputVariablesBuilder.add(newVar);
+                    equiJoinClausesBuilder.add(new JoinNode.EquiJoinClause(var, newVar));
+                    assignmentsBuilder.put(newVar, assignments.get(var));
+                }
+                else {
+                    outputVariablesBuilder.add(var);
+                    assignmentsBuilder.put(var, assignments.get(var));
+                }
+            }
+
+            List<VariableReferenceExpression> newOutputVariables = outputVariablesBuilder.build();
+            ImmutableMap<VariableReferenceExpression, ColumnHandle> newAssignments = assignmentsBuilder.build();
+
+            System.out.println("Old assignments: " + assignments);
+            System.out.println("New assignments: " + newAssignments);
+
+            TableScanNode newScanNode = new TableScanNode(
```

Comment:
This might be good to move to PlannerUtils so we can use it for other optimizations as well. Also we may have to check if the table scan is cloneable. For example if it's a thrift connector, we shouldn't be doing this